source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,345 @@
+/*******************************************************************************
+
+    Contains MerkleTree that makes binary tree of transactions.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.MerkleTree;
+
+import agora.common.Data;
+import agora.common.Hash;
+import agora.common.Transaction;
+
+import std.algorithm;
+import std.conv;
+import std.math;
+
+
+/***************************************************************************
+
+    Create MerkleTree
+
+    Params:
+        txs = Array of `Transaction`
+
+    Returns:
+        Return root node.
+
+***************************************************************************/
+public MerkleNode* MakeMerkleTree (ref Transaction[] txs)
+{
+    MerkleNode[] nodes = new MerkleNode[](txs.length);
+    foreach (size_t idx, ref node; nodes)
```

Comment:
I think you can write this as:

```D
MarkleNode[] nodes;
nodes.reserve(txs.length);
foreach (ref tx; txs)
    nodes ~= MerkleNode(null, null, &tx);
```

Reply:
The current version is more efficient though as it does not go into the GC

Reply:
ok @Geod24 is right